File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> umap;

        int i=0;
        int j=0;
        int maxlen=0;
        while(i<s.size()){
            umap[s[i]]++;
            while(umap[s[i]]>1){
                umap[s[j]]--;
                j++;
            }
            maxlen=max(maxlen,i-j+1);
            i++;
        }
        return maxlen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> umap=new HashMap<>();

        int left=0;
        int right=0;
        int maxlen=0;
        while(right<s.length()){
            char curentchar=s.charAt(right);
            umap.put(curentchar, umap.getOrDefault(curentchar, 0) + 1);
            while(umap.get(curentchar)>1){
                char ch=s.charAt(left);
                umap.put(ch,umap.get(ch)-1);
                left++;
            }
            maxlen=Math.max(maxlen,right-left+1);
            right++;
        }
        return maxlen;
    }
}